### PR TITLE
Fixed an issue where EXPLAIN and EXPLAIN ANALYZE not working if blank line is there in the SQL query.

### PR DESCRIPTION
**Pull request by asheshv**
## Summary

- Use the syntax tree as the primary check for detecting when a blank-line boundary cuts through a SQL statement
- If a Statement node straddles `startPos` (starts before AND ends after), the blank line split a clause and the query is expanded past the blank line
- A keyword list (`STATEMENT_STARTERS`) is retained as a secondary guard for the case where the parser merges semicolon-less queries into one Statement node
- Fixed a Lezer boundary bug: `tree.iterate` is inclusive at boundaries, added `node.to > startPos` check to prevent false positives
- Added 19 new tests covering EXPLAIN, negative cases, parser diagnostics, and boundary edge cases
- All 843 JS tests pass (140 suites)

Supersedes #9670

Refs #2431, #6308

## Test plan

- [x] All existing CodeMirror editor tests pass (15 original tests)
- [x] EXPLAIN ANALYZE with blank lines between clauses returns full query
- [x] EXPLAIN with cursor on FROM clause returns full query including EXPLAIN
- [x] Single-line EXPLAIN without blank lines works (regression guard)
- [x] Multiple consecutive blank lines between clauses expand correctly
- [x] Two separate queries with semicolons are NOT merged (negative)
- [x] Two separate queries without semicolons are NOT merged (negative)
- [x] Comment block between queries does NOT expand into adjacent query (negative)
- [x] First query is NOT expanded into second query (negative)
- [x] Cursor on blank line returns nearest query above
- [x] Parser diagnostic: Lezer boundary inclusivity verified empirically
- [x] Query at previous Statement boundary is not incorrectly expanded
- [x] Full JS test suite: 843 tests, 140 suites, 0 failures

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved query extraction to correctly return full PostgreSQL statements across blank lines, comments, and complex clause boundaries; handles EXPLAIN/EXPLAIN ANALYZE cases.

* **Refactor**
  * Internal extraction logic reorganized for more reliable boundary detection and expansion behavior.

* **Tests**
  * Expanded test coverage for cursor-to-query resolution, statement boundary scenarios, and edge cases to prevent incorrect merging of separate queries.

### CODE REVIEW

**Comment by asheshv** [comment]
Addressed all CodeRabbit findings in 438efea:

**Major fix — asymmetric expansion check:**
- Added `statementExtendsPastEnd` check alongside `statementStraddlesStart` — now detects Statements crossing either boundary
- Introduced `WRAPPER_STARTERS` set (`EXPLAIN`, `ANALYZE`, `WITH`) that always force expansion when the Statement extends past the extracted range
- Cursor on EXPLAIN keyword now correctly returns the full query across blank lines

**Nitpicks fixed:**
- Simplified split regex: `/[\s\n\r(;]+/` → `/[\s(;]+/` (`\s` already includes `\n`/`\r`)
- Strengthened weak parser diagnostic test: pinned exact Statement count with `expect([1, 2]).toContain(stmts.length)`

**New tests added:**
- `EXPLAIN with cursor on EXPLAIN keyword returns full query across blank lines`
- `EXPLAIN ANALYZE with cursor on SELECT returns full query`

All 845 JS tests pass (140 suites, 36 CodeMirror-specific).

File: web/pgadmin/static/js/components/ReactCodeMirror/CustomEditorView.js [hunk 1]
```diff
@@ -50,136 +64,199 @@ export default class CustomEditorView extends EditorView {
...
+    // keywords like FROM/WHERE, identifiers, expressions, etc.)
+    // means the blank line split a statement — expand.
+
+    // Comment-only blocks are self-contained — don't expand.
+    if (query.startsWith('--') || query.startsWith('/*')) return false;
+
+    const firstWord = query.split(/[\s\n\r(;]+/)[0].toUpperCase();
+
+    return !STATEMENT_STARTERS.has(firstWord);
+  }
```
[the remark this answers is not in the document]
Comment (asheshv) [reply web/pgadmin/static/js/components/ReactCodeMirror/CustomEditorView.js:128]: Fixed in 438efea.

1. Added `statementExtendsPastEnd` check — `_needsExpansion` now detects Statements crossing either boundary (`node.from < startPos && node.to > startPos` OR `node.from < endPos && node.to > endPos`), not just startPos.
2. Introduced `WRAPPER_STARTERS` set (`EXPLAIN`, `ANALYZE`, `WITH`) — these always force expansion when the Statement extends past endPos, since they can never stand alone.
3. Simplified split regex: `\s` already includes `\n`/`\r`.
4. Added two new tests verifying cursor on EXPLAIN keyword (pos 4) and cursor on SELECT within EXPLAIN ANALYZE (pos 16) both return the full query.

File: web/regression/javascript/components/CodeMirrorCustomEditor.spec.js [hunk 2]
```diff
@@ -115,4 +116,194 @@ describe('CodeMirrorCustomEditorView', ()=>{
...
+    // Cursor at WHERE clause (position 42), should return full query including EXPLAIN
+    expect(editor.getQueryAt(42)).toEqual({'value': query, 'from': 0, 'to': query.length});
+  });
+
+  it('EXPLAIN with cursor on FROM clause',()=>{
+    const query = 'EXPLAIN SELECT *\n\nFROM pg_class\n\nWHERE oid > 1000;';
+    cmRerender({value: query});
+    // Cursor at FROM (position 18), should return full query including EXPLAIN
+    expect(editor.getQueryAt(18)).toEqual({'value': query, 'from': 0, 'to': query.length});
+  });
```
[the remark this answers is not in the document]
Comment (asheshv) [reply web/regression/javascript/components/CodeMirrorCustomEditor.spec.js]: Fixed in 438efea. Added both suggested tests:

- `EXPLAIN with cursor on EXPLAIN keyword returns full query across blank lines` — cursor at position 4, verifies expansion past blank lines
- `EXPLAIN ANALYZE with cursor on SELECT returns full query` — cursor at position 16 (on SELECT, before any blank line), verifies full query including EXPLAIN ANALYZE prefix

These now pass thanks to the `statementExtendsPastEnd` + `WRAPPER_STARTERS` fix in the same commit.

**Comment by asheshv** [comment]
Responding to CodeRabbit nitpick comments from the review:

**Nitpick 1 — Split STATEMENT_STARTERS into starters + wrappers:**
Done. Added `WRAPPER_STARTERS` (`EXPLAIN`, `ANALYZE`, `WITH`) as a separate `Set`. These always force expansion when the Statement extends past `endPos`. `STATEMENT_STARTERS` retains the full list for the straddle-start case (discriminating new queries from continuations).

**Nitpick 2 — Regex redundancy and comment detection:**
Fixed. Regex simplified from `/[\s\n\r(;]+/` to `/[\s(;]+/`. The `query.startsWith('--')` check operates on already-trimmed text (line 82: `.trim()`), so leading whitespace is already removed. This lightweight check is sufficient because `_findQueryBoundaries` already segments comments into separate ranges — a fully-commented range reaching `_needsExpansion` is expected to start with `--` or `/*` after trim.

**Nitpick 3 — Weak parser diagnostic test:**
Strengthened. Changed from `expect(stmts.length).toBeGreaterThanOrEqual(1)` to `expect([1, 2]).toContain(stmts.length)` with an explicit check on merged span if parser produces 1 Statement. This pins down the current parser behavior so a parser update is noticed.

**Comment by dpage** [comment]
Merged to `master` as 397be4e70 — thanks @asheshv.

Squashed the three commits into one (you preserved as author) and added a release-notes entry for 9.9. Used `Issue #9875` in the changelog since there's no separate tracking issue (the PR's referenced #2431/#6308 are unrelated, and #9670 is the superseded PR) — GitHub redirects the /issues/9875 link to the PR.

Verified before push: 36/36 JS tests pass against the rebased code and ESLint is clean.

Closing manually since the squash changes the SHA.
